src/gsym/linetab.rs

```diff
@@ -0,0 +1,114 @@
+//! Opcode runner of GSYM line table.
+//!
+//! See https://github.com/YtnbFirewings/gsym
+use super::types::{LineTableHeader, LineTableRow};
+use crate::tools::{decode_leb128, decode_leb128_s};
+
+const DBG_END_SEQUENCE: u8 = 0x00; // End of the line table
+const DBG_SET_FILE: u8 = 0x01; // Set LineTableRow.file_idx, don't push a row
+const DBG_ADVANCE_PC: u8 = 0x02; // Increment LineTableRow.address, and push a row
+const DBG_ADVANCE_LINE: u8 = 0x03; // Set LineTableRow.file_line, don't push a row
+const DBG_FIRST_SPECIAL: u8 = 0x04; // All special opcodes push a row.
+
+pub enum RunResult {
+    /// Run the operator successfully.
+    Ok(usize),
+    /// This operator creates a new row.
+    NewRow(usize),
+    /// The end of the program (the operator stream.)
+    End,
+    /// Fails to run the operator at the position.
+    Err,
+}
+
+/// Create a LineTableRow to use as the states of a line table virtual
+/// machine.
+///
+/// The returned LineTableRow can be passed to [`run_op()`] as `ctx`.
+///
+/// # Arguments
+///
+/// * `lthdr` - is a LineTableHeader returned by [`parse_line_table_header()`].
+/// * `symaddr` - the address of the symbol that `lthdr` belongs to.
+#[inline(always)]
+pub fn line_table_row_from(lthdr: &LineTableHeader, symaddr: u64) -> LineTableRow {
+    LineTableRow {
+        address: symaddr,
+        file_idx: 0,
+        file_line: lthdr.first_line,
+    }
+}
+
+/// Run a GSYM line table operator/instruction in the buffer.
+///
+/// # Arguments
+///
+/// * `ctx` - a line table row to present the current states of the
+///           virtual machine.  [`line_table_row_from()`] can create a
+///           LineTableRow to keep the states of a virtual machine.
+/// * `ltbl_hdr` - is a LineTableHeader.
+/// * `ops` - is the buffer of the operators following the
+///           LineTableHeader in a GSYM file.
+/// * `pc` - is the program counter of the virtual machine.
+///
+/// Return a RunResult. `Ok()` and `NewRow()`will return the size of
+/// this instruction.  So, the caller should adjust the value of `pc`
+/// according to the value returned.
+#[inline(always)]
```

Comment:
So this is a 50+ line function. To the degree I can tell it makes no sense to inline it. This annotation actively conveys a need for the function to be inlined. Perhaps (and that is a big perhaps), if there is a single caller it could make sense (though the need part is still open). However, nobody is going to enforce that across module boundaries. Can we please stop annotating stuff with #[inline(always)] without any kind of justification? The only rationale I can conceive of is for proven-to-be-performance-relevant stuff. For everything else the compiler should be trusted to make the right call and #[inline] is sufficient. In reality, even that is probably only relevant on the crate's public API surface and not internal functions, especially with availability (and ease of access) of LTO.

Reply:
As you said, this function will have only one caller, which is called in a loop to run over instructions.  That means it will be called repetitively in a period.  Very often, code like this (interpreters) will be inlined manually, writing just in the loop.  I think it is a proper reason to inline it.

However, I can move the loop of the caller to a function in this module to avoid the concern of crossing modules.

I don't know LTO can help on this to determine if a function of such size is worth being inlined.  PGO may find this function is called in the loop intensively.